### PR TITLE
[POAE7-2438] Support Case/IF Function for Arrow Format

### PR DESCRIPTION
**Pull request by haiweizh**
### What changes were proposed in this pull request?

Support Case/IF Function for Arrow Format.
Sub task of [POAE7-2422](https://jira.devtools.intel.com/browse/POAE7-2422).
  1. Case when/ IF then function without Cast Function.
  2. Case when/ IF then function without String Function.
  3. Case when/ IF then function without Agg Function.

### Why are the changes needed?

Support Case/IF Function for Arrow Format.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Uts.

### Which label does this PR belong to?

cider,test

### CODE REVIEW

File: cider/exec/template/CaseIR.cpp [hunk 1]
```diff
@@ -124,3 +124,111 @@ llvm::Value* CodeGenerator::codegenCase(const Analyzer::CaseExpr* case_expr,
...
+                fmt::format("case type is {}, not support", case_ti.get_type()));
+  }
+  CHECK(case_llvm_type);
+  const auto& else_ti = case_expr->get_else_expr()->get_type_info();
+  CHECK_EQ(else_ti.get_type(), case_ti.get_type());
+  std::unique_ptr<CodegenColValues> case_val =
+      codegenCaseExpr(case_expr, case_llvm_type, is_real_str, co);
+  if (is_real_str) {
+    // FIXME(haiwei): [POAE7-2457] ,blocking by [POAE7-2415]
+    CIDER_THROW(CiderCompileException,
```
Comment (spevenhe) [inline cider/exec/template/CaseIR.cpp:162]: For String, the origin return is a vector contains multi values.  Will return a single ```std::unique_ptr<CodegenColValues>``` be not compatible with ```std::vector<llvm::Value*>```?

File: cider/exec/template/CaseIR.cpp [hunk 2]
```diff
@@ -124,3 +124,111 @@ llvm::Value* CodeGenerator::codegenCase(const Analyzer::CaseExpr* case_expr,
...
+  cgen_state_->ir_builder_.CreateBr(end_bb);
+  cgen_state_->ir_builder_.SetInsertPoint(end_bb);
+  auto then_phi =
+      cgen_state_->ir_builder_.CreatePHI(case_llvm_type, expr_pair_list.size() + 1);
+  CHECK_EQ(then_bbs.size(), then_lvs.size());
+  for (size_t i = 0; i < then_bbs.size(); ++i) {
+    then_phi->addIncoming(then_lvs[i], then_bbs[i]);
+  }
+  then_phi->addIncoming(else_lv, else_bb);
+  return std::make_unique<FixedSizeColValues>(then_phi, null_value);
```
Comment (yma11) [inline cider/exec/template/CaseIR.cpp:233]: Will this `null_value` only be determined by `if_expr`?@spevenhe

File: cider/exec/template/CaseIR.cpp (hunk 1 above)
Comment (haiweizh) [reply cider/exec/template/CaseIR.cpp:162]: This has been discussed with Jiang, will inherit CodegenColValues to contains std::vector<llvm::Value*> in [POAE7-2457] Support Case/IF Function for Arrow Format with String Function after [POAE7-2415] is supported.

File: cider/exec/template/CaseIR.cpp (hunk 2 above)
Comment (haiweizh) [reply cider/exec/template/CaseIR.cpp:233]: This case has been test, but now our test frame will only check row num and column num.

Comment (haiweizh) [reply cider/exec/template/CaseIR.cpp:233]: I have try that null_value should be then or else's null_value, but it is wrong.

File: cider-velox/src/CiderJoinBuild.cpp [hunk 3]
```diff
@@ -25,7 +25,7 @@
 
 namespace facebook::velox::plugin {
 
-void CiderJoinBridge::setData(std::vector<RowVectorPtr> data) {
+void CiderJoinBridge::setData(std::vector<VectorPtr> data) {
```
Comment (jikunshang) [inline cider-velox/src/CiderJoinBuild.cpp:28]: is this your code change? seems curious.

Comment (haiweizh) [reply cider-velox/src/CiderJoinBuild.cpp:28]: That is rebase code.

File: cider/exec/template/CodeGenerator.h [hunk 4]
```diff
@@ -217,6 +217,7 @@ class CodeGenerator {
 
   llvm::Value* codegen(const Analyzer::InIntegerSet* expr, const CompilationOptions& co);
 
+  // Deprecating
```
Comment (winningsix) [inline cider/exec/template/CodeGenerator.h:220]: Use TODO instead to call this out. No need to follow previous style.

Comment (haiweizh) [reply cider/exec/template/CodeGenerator.h]: Done.

**Comment by haiweizh** [comment]
Use new arrow test framework to do the if/case-when test.

File: cider/tests/functionality/CiderCaseWhenTest.cpp [hunk 5]
```diff
@@ -82,6 +84,65 @@ class CiderCaseWhenRandomWithNullTestBase : public CiderTestBase {
   }
 };
 
+class CiderArrowCaseWhenSequenceTestBase : public CiderTestBase {
+ public:
+  CiderArrowCaseWhenSequenceTestBase() {
+    table_name_ = "test";
+    create_ddl_ =
+        R"(CREATE TABLE test(col_int INTEGER, col_bigint BIGINT, col_double DOUBLE, col_float FLOAT);)";
```
Comment (jikunshang) [inline cider/tests/functionality/CiderCaseWhenTest.cpp:92]: Please add NOT NULL.

File: cider/exec/template/CaseIR.cpp [hunk 6]
```diff
@@ -124,3 +124,111 @@ llvm::Value* CodeGenerator::codegenCase(const Analyzer::CaseExpr* case_expr,
...
+    cgen_state_->ir_builder_.CreateBr(end_bb);
+    const auto when_bb = llvm::BasicBlock::Create(
+        cgen_state_->context_, "when_case", cgen_state_->current_func_);
+    cgen_state_->ir_builder_.SetInsertPoint(cmp_bb);
+    cgen_state_->ir_builder_.CreateCondBr(when_lv, then_bb, when_bb);
+    cgen_state_->ir_builder_.SetInsertPoint(when_bb);
+  }
+  const auto else_expr = case_expr->get_else_expr();
+  CHECK(else_expr);
+  const auto else_expr_ptr = codegen(else_expr, co, true);
```
Comment (yma11) [inline cider/exec/template/CaseIR.cpp:219]: Should `else_expr_ptr ` also call new codegen API?

File: cider/tests/functionality/CiderCaseWhenTest.cpp (hunk 5 above)
Comment (haiweizh) [reply cider/tests/functionality/CiderCaseWhenTest.cpp]: Done.

File: cider/exec/template/CaseIR.cpp (hunk 2 above)
Comment (yma11) [reply cider/exec/template/CaseIR.cpp:233]: I think the `null_value` should be determined by ` if_expr || then` or ` if_expr || else` but not sure how describe this. You may switch to latest test framework and find out the clear logic.

**Comment by haiweizh** [comment]
PR Ready, Plz help to take a review. @yma11 @jikunshang

File: cider/exec/template/CaseIR.cpp (hunk 2 above)
Comment (spevenhe) [reply cider/exec/template/CaseIR.cpp:233]: The null may be decided in ```columnvar codgen```, so may not be ```then``` or ```else``` 's null_value

File: cider/tests/functionality/CiderCaseWhenTest.cpp [hunk 7]
```diff
@@ -272,6 +354,93 @@ CASE_WHEN_AGG_TEST(CiderCaseWhenSequenceTestBase, caseWhenAggTest);
...
+#define CASE_WHEN_ARROW_TEST(TEST_CLASS, UNIT_NAME)                                      \
+  TEST_F(TEST_CLASS, UNIT_NAME) {                                                        \
+    assertQueryArrow(                                                                    \
+        "SELECT col_int, CASE WHEN col_int > 3 THEN 1 ELSE 0 END FROM test");            \
+    assertQueryArrow(                                                                    \
+        "SELECT col_double, CASE WHEN col_int > 5 THEN 4 ELSE 3"                         \
+        " END FROM test");                                                               \
+    assertQueryArrowIgnoreOrder(                                                         \
+        "SELECT col_int, CASE WHEN col_int > 30 THEN 1 ELSE 0 END FROM test");           \
+    GTEST_SKIP_("FIXME(haiwei): Case when with null value");                             \
```
Comment (jikunshang) [inline cider/tests/functionality/CiderCaseWhenTest.cpp:366]: I remember one `GTEST_SKIP_` will disable all later case?

**Review by jikunshang — APPROVED** [review]

File: cider/tests/functionality/CiderCaseWhenTest.cpp (hunk 7 above)
Comment (haiweizh) [reply cider/tests/functionality/CiderCaseWhenTest.cpp:366]: Only disable the TEST_F's case. Case when with null value will been finished in another sub-task.

**Review by yma11 — APPROVED** [review]
Let's address the null_value in later PR.

**Pull request merged**